06_puppy/mohitnag/store_test.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+type storer struct {
+	suite.Suite
+	store      Storer
+	makeStorer func() Storer
+	puppy      Puppy
+}
+
+func (s *storer) SetupTest() {
+	s.store = s.makeStorer()
+	s.puppy = Puppy{ID: 1, Breed: "dog", Colour: "white", Value: "$2"}
+}
+
+func (s *storer) TestCreatePuppy() {
+	assert := assert.New(s.T())
+	testCases := []struct {
+		Scenario      string
+		Input         Puppy
+		ExpectedError string
+	}{
+		{"Create Puppy", s.puppy, "<nil>"},
+		{"Creating already existing Puppy should fail", s.puppy, "puppy with Id 1 already exists"},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		s.T().Run(tc.Scenario, func(t *testing.T) {
+			err := s.store.CreatePuppy(tc.Input)
+			assert.Equal(tc.ExpectedError, fmt.Sprint(err))
+		})
+	}
+}
+
+func (s *storer) TestReadPuppy() {
+	assert := assert.New(s.T())
+	err := s.store.CreatePuppy(s.puppy)
+	assert.NoError(err)
+	testCases := []struct {
+		Scenario      string
+		ID            uint32
+		Expected      Puppy
+		ExpectedError string
+	}{
+		{"Read already existing Puppy", 1, s.puppy, "<nil>"},
+		{"Reading a non-existing Puppy should fail", 32, Puppy{}, "puppy with Id 32 does not exists"},
+	}
+	for _, tc := range testCases {
+		tc := tc
+		s.T().Run(tc.Scenario, func(t *testing.T) {
+			puppy, err := s.store.ReadPuppy(tc.ID)
+			assert.Equal(tc.Expected, puppy)
+			assert.Equal(tc.ExpectedError, fmt.Sprint(err))
+		})
+	}
+
+}
+
+func (s *storer) TestUpdatePuppy() {
+	assert := assert.New(s.T())
+	err := s.store.CreatePuppy(s.puppy)
```

Comment:
See how on all but one of these tests your first step is to create a new puppy based off the puppy stored in the storer. What you could do here is move this logic into the `SetupTest` function. This will remove the need to duplicate the logic across these three tests, as well as remove the need to store the puppy inside the store.

Reply:
Done